In [11]:
!pip install numpy-financial

In [5]:
import numpy_financial as npf
import numpy as np




# IRR

In [6]:
cf=np.array([-200,20,50,70,100,50])
r=0.06

In [7]:
npf.npv(r,cf)

38.71337130837991

In [8]:
npf.irr(cf)

0.1190693988331708

In [10]:
npf.irr(cf)>r

True

# YTM

1OY BOND 5% COUPON RATE, TODAY PRICE=107.5

In [23]:
cf=[5]*9

In [24]:
cf[-1]=cf[-1]+100

In [25]:
cf.insert(0,-107.5)

In [26]:
cf

[-107.5, 5, 5, 5, 5, 5, 5, 5, 5, 105]

In [27]:
npf.irr(cf)

0.039916847253554044

# ANNUITIES 

SAVE 2000 PA AT 3% INTEREST FOR 25 YEARS

In [34]:
pv=0
n=25
r=0.03
cf=2000

In [35]:
fv= npf.fv(rate=r,nper=n,pmt=cf,pv=pv)
fv

-72918.52864361441

change initial pv to 10000

In [36]:
pv=10000

In [37]:
fv=npf.fv(rate=r,nper=n,pmt=cf,pv=pv)
fv

-93856.30794015658

this shows that fv of 10k is around 21k 
to check we can calculate fv of 10k only

In [41]:
val=10000*(1+r)**n

In [42]:
val

20937.77929654216

In [43]:
val1=npf.fv(rate=r,nper=n,pmt=0,pv=pv)
val1

-20937.77929654216

ANNUITY CASE:
    FUNDING=100,000
    5K PAYOUTS FOR AT 3% INTEREST
    VALUE OF REMAINING FV AFTER 25 YEARS

In [65]:
pv=-100000
cf=-5000
r=0.03

intuition: what is the total time period for us to get all our money back (will use it to calculate fv):

In [147]:
pv=-100000
cf=-5000
r=0.03

In [141]:
npf.nper(rate=r,pmt=cf,pv=100000,fv=0)

array(30.99891276)

30.998 is total payback period, will use to calculate total fv

In [148]:
npf.fv(rate=r,nper=30.998,pmt=cf,pv=pv)


499982.0135821306

first understand the mechanics of payout phase:

In [45]:
pv*(1+r)-cf

98000.0

that is what the contract is worth after a raise of 3% interest and a payout of 5k at end of year1

In [47]:
(pv*(1+r)-cf)*(1+r)-cf

95940.0

this is the contract worth after second year payment

In [48]:
pv=-100000
cf=5000
n=25
r=0.03

In [49]:
fv=npf.fv(rate=r,nper=25,pmt=cf,pv=pv)
fv

27081.47135638559

that is the worth of contract after 25 years of 5k payments 

LETS ASSUME YOU WANT A 100K WORTH CONTRACT AT END OF 25 YEARS, YOU HAVE 3% INTEREST RATE AND 0 INVESTED, SO U NEED TO CALCULATE HOW MUCH YOU NEED TO GIVE ANNUALLY 

In [56]:
pv=0
fv=100000
r=0.03
n=25

In [57]:
cf=npf.pmt(rate=r,nper=n,pv=pv,fv=fv)
cf

-2742.7871039127763

to confirm lets calculate fv based on the above cf

In [58]:
npf.fv(rate=r,nper=25,pmt=cf,pv=0)

99999.99999999999

N PAYMENTS OF 2.5K AT 3% INTEREST TO GET FV OF 100K

In [97]:
pv=0
r=0.03
cf=-2500
fv=100000

In [98]:
npf.nper(rate=r,pmt=cf,pv=pv,fv=fv)

array(26.67419857)

In [100]:
npf.fv(rate=r,nper=26.67419857,pmt=cf,pv=0)

100000.00002164196

AT RETIREMENT YOUR CONTRACT IS ANNUITY WORTH X USD. AT 3% INTEREST YOU RECEIVE 5K PAYOUTS AND FV AT 25 YEARS IS 20K, FIND PV

In [102]:
fv=20000
r=0.03
n=25
cf=5000

In [105]:
pv=npf.pv(rate=r,nper=n,pmt=cf,fv=fv)
pv

-96617.84984162332

this means we hae to pay 96618 which returns 5k at 3% annually

In [118]:
0.03/12

0.0025